numba/runtime/_nrt_python.c

```diff
@@ -0,0 +1,539 @@
+    PyGILState_Release(gstate);     /* release the GIL */
+}
+
+
+static
+MemInfo* meminfo_new_from_pyobject(void *data, PyObject *ownerobj) {
+    size_t dummy_size = 0;
+    Py_INCREF(ownerobj);
+    return NRT_MemInfo_new(data, dummy_size, pyobject_dtor, ownerobj);
+}
+
+
+/*
+ * Create a new MemInfo with a owner PyObject
+ */
+static
+PyObject*
+meminfo_new(PyObject *self, PyObject *args) {
+    unsigned PY_LONG_LONG addr_data;
+    PyObject* ownerobj;
+    MemInfo *mi;
+    if (!PyArg_ParseTuple(args, "KO", &addr_data, &ownerobj)) {
+        return NULL;
+    }
+    mi = meminfo_new_from_pyobject((void*)addr_data, ownerobj);
+    return PyLong_FromVoidPtr(mi);
+}
+
+/*
+ * Create a new MemInfo with a new NRT allocation
+ */
+static
+PyObject*
+meminfo_alloc(PyObject *self, PyObject *args) {
+    MemInfo *mi;
+    Py_ssize_t size;
+    if (!PyArg_ParseTuple(args, "K", &size)) {
+        return NULL;
+    }
+    mi = NRT_MemInfo_alloc(size);
+    return PyLong_FromVoidPtr(mi);
+}
+
+/*
+ * Like meminfo_alloc but set memory to zero after allocation and before
+ * deallocation.
+ */
+static
+PyObject*
+meminfo_alloc_safe(PyObject *self, PyObject *args) {
+    MemInfo *mi;
+    Py_ssize_t size;
+    if (!PyArg_ParseTuple(args, "K", &size)) {
+        return NULL;
+    }
+    mi = NRT_MemInfo_alloc_safe(size);
+    return PyLong_FromVoidPtr(mi);
+}
+
+typedef struct {
+    PyObject_HEAD
+    MemInfo *meminfo;
+    int      defer;
+} MemInfoObject;
+
+static
+int MemInfo_init(MemInfoObject *self, PyObject *args, PyObject *kwds) {
+    static char *keywords[] = {"ptr", NULL};
+    unsigned PY_LONG_LONG raw_ptr;
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "K", keywords, &raw_ptr)) {
+        return -1;
+    }
+    self->meminfo = (MemInfo*)raw_ptr;
+    self->defer = 0;
+    NRT_MemInfo_acquire(self->meminfo);
+    return 0;
+}
+
+int MemInfo_getbuffer(PyObject *exporter, Py_buffer *view, int flags) {
+    Py_ssize_t len;
+    void *buf;
+    int readonly = 0;
+
+    MemInfoObject *miobj = (MemInfoObject*)exporter;
+    MemInfo *mi = miobj->meminfo;
+
+    buf = NRT_MemInfo_data(mi);
+    len = NRT_MemInfo_size(mi);
+    return PyBuffer_FillInfo(view, exporter, buf, len, readonly, flags);
+}
+
+Py_ssize_t MemInfo_rdwrbufferproc(PyObject *self, Py_ssize_t segment,
+                                  void **ptrptr)
+{
+    MemInfoObject *mio = (MemInfoObject *)self;
+    MemInfo *mi = mio->meminfo;
+    if (segment != 0) {
+        PyErr_SetString(PyExc_TypeError, "MemInfo only has 1 segment");
+        return -1;
+    }
+    *ptrptr = NRT_MemInfo_data(mi);
+    return NRT_MemInfo_size(mi);
+}
+
+Py_ssize_t MemInfo_segcountproc(PyObject *self, Py_ssize_t *lenp) {
+    MemInfoObject *mio = (MemInfoObject *)self;
+    MemInfo *mi = mio->meminfo;
+    if (lenp) {
+        *lenp = NRT_MemInfo_size(mi);
+    }
+    return 1;
+}
+
+#if (PY_MAJOR_VERSION < 3)
+static PyBufferProcs MemInfo_bufferProcs = {MemInfo_rdwrbufferproc,
```

Comment:
Instead of that, you can implement the new buffer protocol on Python 2. Enable the Py_TPFLAGS_HAVE_NEWBUFFER and see here for an example:
https://hg.python.org/cpython/file/4bf210f59ac6/Objects/bytearrayobject.c#l2800